### PR TITLE
[RFC] introduce `dvc tag`

### PR DESCRIPTION
**Pull request by efiop**
Initial implementation with basic `dvc tag add/list/remove` support and support for `dvc checkout stage.dvc@mytag`.

### CODE REVIEW

File: dvc/command/tag.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                if tags:
+                    logger.info(yaml.dump(tags, default_flow_style=False))
+            except DvcException:
+                logger.error("failed list tags")
+                return 1
+        return 0
+
+
+def add_parser(subparsers, parent_parser):
+    TAG_HELP = "A set of commands to manage dvc tags."
```
Comment (shcheklein) [inline dvc/command/tag.py]: minor: let's use uppercase DVC everywhere.

File: dvc/command/tag.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    tag_parser = subparsers.add_parser(
+        "tag", parents=[parent_parser], description=TAG_HELP, add_help=False
```
Comment (shcheklein) [inline dvc/command/tag.py]: minor: add documentation link

File: dvc/output/local.py [hunk 3]
```diff
@@ -155,12 +157,7 @@ def save(self):
 
     @property
     def dir_cache(self):
-        if self.checksum not in self._dir_cache.keys():
-            self._dir_cache[
-                self.checksum
-            ] = self.repo.cache.local.load_dir_cache(self.checksum)
-
-        return self._dir_cache[self.checksum]
+        return self.repo.cache.local.load_dir_cache(self.checksum)
```
Comment (shcheklein) [inline dvc/output/local.py]: why do we remove this optimization(?), is it not relevant for outputs in general?

File: dvc/repo/checkout.py [hunk 4]
```diff
@@ -29,15 +29,14 @@ def checkout(self, target=None, with_deps=False, force=False, recursive=False):
...
-            )
-    else:
-        stages = self.active_stages(target)
+    try:
+        stages = self.collect(target, with_deps=with_deps, recursive=recursive)
+    except (StageFileDoesNotExistError, StageFileBadNameError) as exc:
+        if not target:
+            raise
+        raise DvcException(
+            str(exc) + " Did you mean 'git checkout {}'?".format(target)
```
Comment (shcheklein) [inline dvc/repo/checkout.py]: is better to just pass `cause=exc` everywhere and let the top level wrapper handler print it properly?

**Comment by shcheklein** [comment]
Looks awesome! Initial questions that come to my mind.

1. We should probably support `--all-branches` at least for `dvc tag list`, `dvc checkout tag` etc. The same will be true for the `pkg install` we will need to specify the branch somehow.
2. With this interface how do I get the git commit with the pipeline that corresponds to the model built? Should we be saving more context and give some interface to go back if needed?

File: dvc/output/local.py (hunk 3 above)
Comment (efiop) [reply dvc/output/local.py]: I didn't remove it, I just moved it to RemoteLOCAL class, where it is more suitable.

File: dvc/repo/checkout.py (hunk 4 above)
Comment (efiop) [reply dvc/repo/checkout.py]: Yeah, that would also look good enough. I'll adjust it. Thanks! :slightly_smiling_face:

**Comment by efiop** [comment]
@shcheklein Thanks for the review! :slightly_smiling_face: 

> We should probably support --all-branches at least for dvc tag list, dvc checkout tag etc. The same will be true for the pkg install we will need to specify the branch somehow.

+1 for `dvc tag list`. Not sure how `dvc checkout tag` would work with `--all-branches` though. But looks like you are talking about git branch/reference support for `dvc checkout`, right?

> With this interface how do I get the git commit with the pipeline that corresponds to the model built? Should we be saving more context and give some interface to go back if needed?

No way to do that through the interface itself, nor do I think that it should be a part of `dvc tag` or any `dvc` command. A user would just need to utilize git tags, as usual, for that. Otherwise, we would have to record git references/tags in our dvc files, which is a bad idea, and better to be left to git itself.

**Comment by shcheklein** [comment]
>  But looks like you are talking about git branch/reference support for dvc checkout, right?

That's exactly right. I should be able to pull any names dataset/model easily to my working space, to an external project, to a production machine. So, we need to think about the naming scheme when ppl are using branches (I'm not sure btw if tags are relevant in this case).

> No way to do that through the interface itself, nor do I think that it should be a part of `dvc tag` or any `dvc` command. A user would just need to utilize git tags, as usual, for that. Otherwise, we would have to record git references/tags in our dvc files, which is a bad idea, and better to be left to git itself.

What are thoughts @dmpetrov ? To some extent it breaks the reproducibility and experimentation management idea that we always should an easy access to the full context of any artifact built.

**Comment by efiop** [comment]
> That's exactly right. I should be able to pull any names dataset/model easily to my working space, to an external project, to a production machine. So, we need to think about the naming scheme when ppl are using branches (I'm not sure btw if tags are relevant in this case).

Agreed. This is not in the scope of initial implementation though. `dvc tag` is hidden from help for now, we can still work on it without being worried about backward compatibility.

**Pull request merged**